cmd/queuebench/main.go

```diff
@@ -0,0 +1,209 @@
+// under the License.
+
+package main
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+
+	apmqueue "github.com/elastic/apm-queue"
+	"github.com/elastic/apm-queue/kafka"
+)
+
+const namespace = "queuebench"
+
+var totalproduced = 0
+var totalconsumed = 0
+
+func main() {
+	// NOTE: intercept any panic and print a nice terminate gracefully
+	// This allows using log.Panic methods in main; those function
+	// trigger deferred functions whereas log.Fatal don't.
+	defer func() {
+		if r := recover(); r != nil {
+			log.Fatal(r)
+		}
+	}()
+
+	cfg := config{}
+
+	cfg.Parse()
+
+	fmt.Printf("%+v\n", cfg)
+
+	log.Println("prep logger")
+	var logger *zap.Logger
+	var err error
+	if cfg.verbose {
+		logger, err = zap.NewDevelopment()
+		if err != nil {
+			log.Fatalf("cannot create zap logger: %s", err)
+		}
+	} else {
+		logger = zap.NewNop()
+	}
+
+	log.Println("prep MeterProvider")
+	rdr := sdkmetric.NewManualReader()
+	mp := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(rdr),
+	)
+
+	ctx := context.Background()
+	run := time.Now().Unix()
+	log.Printf("running bench run: %d", run)
+
+	bench := bench{
+		Brokers:         []string{cfg.broker},
+		ConsumerGroupID: fmt.Sprintf("queuebench-%d", run),
+		Logger:          logger,
+		Partitions:      cfg.partitions,
+		TopicNamespace:  namespace,
+		Topics: []apmqueue.Topic{
+			apmqueue.Topic(fmt.Sprintf("run-%d", run)),
+		},
+
+		mp: mp,
+		tp: trace.NewNoopTracerProvider(),
+	}
+
+	log.Println("running benchmark setup")
+	if err = bench.Setup(ctx); err != nil {
+		log.Panicf("benchmark setup failed: %s", err)
+	}
+	teardown := func() {
+		log.Println("running benchmark teardown")
+		if err := bench.Teardown(ctx); err != nil {
+			log.Panicf("benchmark teardown failed: %s", err)
+		}
+
+	}
+	defer teardown()
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		for range c {
+			teardown()
+		}
+	}()
+
+	start := time.Now()
+	log.Println("==> running benchmark")
+
+	log.Println("start consuming")
+	var consumptionduration time.Duration
+	go func() {
+		consumptionstart := time.Now()
+		if err := bench.c.Run(ctx); err != nil {
+			log.Panicf("consumer run ended with an error: %s", err)
+		}
+		consumptionduration = time.Since(consumptionstart)
+	}()
+
+	log.Printf("start producing, will produce for %s", cfg.duration)
+	productionstart := time.Now()
+	if err := produce(ctx, bench.p, bench.Topics[0], cfg.eventSize, cfg.duration); err != nil {
+		log.Panicf("error while producing records: %s", err)
+	}
+	productionduration := time.Since(productionstart)
+
+	log.Println("production ended")
+
+	log.Println("waiting for consumer to fetch all records")
+	ticker := time.NewTicker(100 * time.Millisecond)
+	timer := time.NewTimer(cfg.timeout)
```

Comment:
I haven't used defer in this case as I wanted to stop both in the proximity of the for loop instead of at the end of `main`, but doesn't make much of a difference and `defer` is more idiomatic so :+1: